File: python/packages/ai/teams/ai/models/openai_function.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Awaitable, Callable, Dict, Optional, Union
from dataclasses_json import DataClassJsonMixin, dataclass_json


@dataclass_json
@dataclass
class OpenAIFunction(DataClassJsonMixin):
    """
    Function spec that adheres to OpenAI function calling.
    """

    name: str
    """
    Name of the function to be called.
    """

    description: Optional[str]
    """
    Description of what the function does.
    """

    parameters: Optional[Dict[str, Any]]
    """
    Parameters the function accepts, described as a JSON Schema object.
    """

    handler: Callable[..., Union[str, Awaitable[str]]]
    """
    The function handler, may be asynchoronous, takes in any number of
    arguments and must return a string.
    """

File: js/packages/teams-ai/src/types/OpenAIFunction.ts
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

/**
 * Interface that adheres to OpenAI tools/function calling.
 */
export interface OpenAIFunction {
    /**
     * The function name to call.
     */
    name: string;

    /**
     * The function handler to call. It may be asynchronous, accepts any number of arguments, and must return a string.
     */
    handler: (...args: any[]) => string | Promise<string>;

    /**
     * Optional. Description of the function.
     */
    description?: string;

    /**
     * Parameters the function accepts, described as a JSON object.
     */
    parameters?: Record<string, any>;
}
